### PR TITLE
Upgrade `spdx-tools` to v0.8.1 #3455

### LINKED ISSUE #3455: Upgrade `spdx-tools` to version 0.8
## Short Description
Upgrade `spdx-tools` to v0.8

## Possible Labels

- dependencies

## Select Category

-  [ ] Enhancement
-  [ ] Add License/Copyright
-  [ ] Scan Feature
-  [ ] Packaging
-  [ ] Documentation
-  [x] Expand Support
-  [ ] Other

## **Describe the Update**

Update the `spdx-tools` dependency from `0.7.0rc0` to `0.8.0rc2` or later.

## **How This Feature will help you/your organization**
The new 0.8 version of the `spdx-tools` lays the groundwork for the upcoming SPDX-3.0 implementation in python. We therefore expect broad adoption of the 0.8 version.
Tools that use the 0.8 version of the `spdx-tools` currently can't simultaneously use scancode which depends on the 0.7 version (there can't be two different versions of the same library in a single environment).
This issue currently blocks adoption of the `spdx-tools` in tern, for example.

## **Example/Links if Any**

Pending PR in tern that introduces `spdx-tools` in version 0.8: https://github.com/tern-tools/tern/pull/1233

## **Can you help with this Feature**

I will open a PR that takes care of the upgrade while preserving the current feature set.

### PR DESCRIPTION
**Pull request by armintaenzertng**
As the `spdx-tools` library has been refactored, this requires some code adaption of the current `output_spdx.py` and regeneration of the test files.
Also, the SPDX license list is now handled by the `license-expression` library, which is why we don't need the corresponding code anymore that adapts this list.

Fixes #3455

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁
* [x] Looked for possible updates in documentation and added updates if applicable
* [x] Updated CHANGELOG.rst

### CODE REVIEW

File: src/formattedcode/output_spdx.py [hunk 1]
```diff
@@ -9,29 +9,22 @@
...
-from spdx.utils import SPDXNone
-from spdx.version import Version
 
 from license_expression import Licensing
 from commoncode.cliutils import OUTPUT_GROUP
 from commoncode.cliutils import PluggableCommandLineOption
 from commoncode.fileutils import file_name
 from commoncode.fileutils import parent_directory
 from commoncode.text import python_safe_name
+from spdx_tools.spdx.model import SpdxNoAssertion, Version, CreationInfo, Actor, ActorType, Document, Package, File, \
```
Comment (pombredanne) [inline src/formattedcode/output_spdx.py]: If you do not mind, our convention is one import per line, no multiline imports.

File: src/formattedcode/output_spdx.py [hunk 2]
```diff
@@ -312,63 +272,63 @@ def write_spdx(
...
-                        spdx_license = ExtractedLicense(spdx_id)
-                        spdx_license.name = file_license.short_name
+                    spdx_license = licensing.parse(spdx_id)
+
+                    if is_license_ref:
+                        text = match.get('matched_text')
                         # FIXME: replace this with the licensedb URL
                         comment = (
                             f'See details at https://github.com/nexB/scancode-toolkit'
                             f'/blob/develop/src/licensedcode/data/licenses/{license_key}.yml\n'
```
Comment (pombredanne) [inline src/formattedcode/output_spdx.py]:
```suggestion
                            f'/blob/develop/src/licensedcode/data/licenses/{license_key}.LICENSE\n'
```

File: src/formattedcode/output_spdx.py [hunk 3]
```diff
@@ -312,63 +272,63 @@ def write_spdx(
...
 
-                    if not is_license_ref:
-                        spdx_license = License.from_identifier(spdx_id)
-                    else:
-                        spdx_license = ExtractedLicense(spdx_id)
-                        spdx_license.name = file_license.short_name
+                    spdx_license = licensing.parse(spdx_id)
+
+                    if is_license_ref:
+                        text = match.get('matched_text')
```
Comment (pombredanne) [inline src/formattedcode/output_spdx.py:289]: I think we would want the matched text in most cases, unless this is matched to a full license text for a standard SPDX license. For all other cases (notices, license_ref, etc.) we should get this.

File: src/formattedcode/output_spdx.py [hunk 4]
```diff
@@ -312,63 +272,63 @@ def write_spdx(
...
+                    package.license_info_from_files += [spdx_license]
 
         elif license_matches is None:
             all_files_have_no_license = False
-            file_entry.add_lics(NoAssert())
+            file_entry.license_info_in_file += [SpdxNoAssertion()]
 
         else:
-            file_entry.add_lics(SPDXNone())
+            file_entry.license_info_in_file += [SpdxNone()]
```
Comment (pombredanne) [inline src/formattedcode/output_spdx.py]: Why is this `license_info_in_file` and not `license_info_in_files` or some plural since this is a list of things?

File: src/formattedcode/output_spdx.py [hunk 5]
```diff
@@ -312,63 +272,63 @@ def write_spdx(
...
+                            license_name=file_license.short_name,
+                            comment=comment,
+                        )
+                        doc.extracted_licensing_info += [extracted_license]
 
                     # Add licenses in the order they appear in the file. Maintaining
                     # the order might be useful for provenance purposes.
-                    file_entry.add_lics(spdx_license)
-                    package.add_lics_from_file(spdx_license)
+                    file_entry.license_info_in_file += [spdx_license]
```
Comment (pombredanne) [inline src/formattedcode/output_spdx.py]: Why not use `.append()` here and in other places? Creating and then adding a list to an existing list feels too much.

File: src/formattedcode/output_spdx.py [hunk 6]
```diff
@@ -379,33 +339,29 @@ def write_spdx(
...
-    if len(package.cr_text) == 0:
+    if not package_copyright_texts:
         if all_files_have_no_copyright:
-            package.cr_text = SPDXNone()
+            package.copyright_text = SpdxNone()
         else:
-            package.cr_text = NoAssert()
+            package.copyright_text = SpdxNoAssertion()
     else:
         # Create a text of alphabetically sorted copyright
```
Comment (pombredanne) [inline src/formattedcode/output_spdx.py:369]: IMHO we should not use a set and instead preserve the original order of copyrights: it matters.

File: tests/formattedcode/data/spdx/license_known/expected.tv [hunk 7]
```diff
@@ -1,35 +1,36 @@
-# Document Information
+## Document Information
```
Comment (pombredanne) [inline tests/formattedcode/data/spdx/license_known/expected.tv:1]: Why using a double ## pound?

File: tests/formattedcode/data/spdx/license_known/expected.tv [hunk 8]
```diff
@@ -1,35 +1,36 @@
...
 SPDXVersion: SPDX-2.2
 DataLicense: CC0-1.0
-DocumentNamespace: http://spdx.org/spdxdocs/scan
```
Comment (pombredanne) [inline tests/formattedcode/data/spdx/license_known/expected.tv:4]: Why change the ordering of the fields?

File: tests/formattedcode/data/spdx/license_known/expected.tv [hunk 9]
```diff
@@ -1,35 +1,36 @@
...
-DocumentName: SPDX Document created by ScanCode Toolkit
-LicenseListVersion: 3.20
```
Comment (pombredanne) [inline tests/formattedcode/data/spdx/license_known/expected.tv:6]: Why removing this version? this sounds a big change?

File: tests/formattedcode/data/spdx/license_known/expected.tv [hunk 10]
```diff
@@ -1,35 +1,36 @@
...
+Visit https://github.com/nexB/scancode-toolkit/ for support and download.
+SPDX License List: 3.21</text>
+## Creation Information
+## Package Information
 PackageName: scan
 SPDXID: SPDXRef-001
 PackageDownloadLocation: NOASSERTION
+FilesAnalyzed: True
 PackageVerificationCode: 517c26f04b5a3edaaa0c17bf1a6571953209e209
-PackageLicenseDeclared: NOASSERTION
```
Comment (pombredanne) [inline tests/formattedcode/data/spdx/license_known/expected.tv:20]: Why not keeping the original ordering? Declared before Concluded is the right natural ordering we have until now?

File: tests/formattedcode/data/spdx/license_known/expected.tv [hunk 11]
```diff
@@ -1,35 +1,36 @@
...
 PackageLicenseConcluded: NOASSERTION
 PackageLicenseInfoFromFiles: Apache-2.0
 PackageLicenseInfoFromFiles: CC0-1.0
+PackageLicenseDeclared: NOASSERTION
 PackageCopyrightText: NONE
-# File
+## File Information
```
Comment (pombredanne) [inline tests/formattedcode/data/spdx/license_known/expected.tv:26]: Why the extra # and the trailing "Information"?

File: tests/formattedcode/data/spdx/license_known/expected.rdf [hunk 12]
```diff
@@ -1,117 +1,105 @@
...
+      },
+      "spdx:relationship": [
         {
-          "ns1:Relationship": {
-            "ns1:relationshipType": {
+          "spdx:Relationship": {
+            "spdx:relationshipType": {
               "@rdf:resource": "http://spdx.org/rdf/terms#relationshipType_contains"
             },
-            "ns1:spdxElementId": "SPDXRef-001",
```
Comment (pombredanne) [inline tests/formattedcode/data/spdx/license_known/expected.rdf:97]: Where have the Ids gone? I thought these were must have data?

File: tests/formattedcode/data/spdx/license_ref/expected_with_text.tv [hunk 13]
```diff
@@ -85,10 +83,10 @@ License for ScanCode datasets
 ScanCode includes datasets (e.g. for license detection) that are dedicated
 to the Public Domain using the Creative Commons CC0 1.0 Universal (CC0 1.0)
 Public Domain Dedication: http://creativecommons.org/publicdomain/zero/1.0/</text>
-LicenseID: LicenseRef-scancode-other-permissive
```
Comment (pombredanne) [inline tests/formattedcode/data/spdx/license_ref/expected_with_text.tv:88]: Why change the order? Id, name then comment is the proper order here.

File: tests/formattedcode/data/spdx/license_ref/expected_with_text.tv [hunk 14]
```diff
@@ -85,10 +83,10 @@ License for ScanCode datasets
...
-LicenseName: Other Permissive Licenses
-LicenseComment: <text>See details at https://github.com/nexB/scancode-toolkit/blob/develop/src/licensedcode/data/licenses/other-permissive.yml
+LicenseName: ScanCode acknowledgment
+LicenseComment: <text>See details at https://github.com/nexB/scancode-toolkit/blob/develop/src/licensedcode/data/licenses/scancode-acknowledgment.yml
 </text>
+LicenseID: LicenseRef-scancode-other-permissive
```
Comment (pombredanne) [inline tests/formattedcode/data/spdx/license_ref/expected_with_text.tv:89]: "scancode-acknowledgment" is not "other-permissive" so there is likely a bug here.

File: tests/formattedcode/data/spdx/license_ref/expected_with_text.tv [hunk 15]
```diff
@@ -131,10 +129,10 @@ License for ScanCode datasets
 ScanCode includes datasets (e.g. for license detection) that are dedicated
 to the Public Domain using the Creative Commons CC0 1.0 Universal (CC0 1.0)
 Public Domain Dedication: http://creativecommons.org/publicdomain/zero/1.0/</text>
-LicenseID: LicenseRef-scancode-scancode-acknowledgment
-LicenseName: ScanCode generated data acknowledgment
-LicenseComment: <text>See details at https://github.com/nexB/scancode-toolkit/blob/develop/src/licensedcode/data/licenses/scancode-acknowledgment.yml
+LicenseName: Other Permissive Licenses
+LicenseComment: <text>See details at https://github.com/nexB/scancode-toolkit/blob/develop/src/licensedcode/data/licenses/other-permissive.yml
 </text>
+LicenseID: LicenseRef-scancode-other-copyleft
```
Comment (pombredanne) [inline tests/formattedcode/data/spdx/license_ref/expected_with_text.tv:135]: Same nit as above

File: tests/formattedcode/data/spdx/license_ref/expected_with_text.tv [hunk 16]
```diff
@@ -176,4 +174,7 @@ License for ScanCode datasets
 =============================
 ScanCode includes datasets (e.g. for license detection) that are dedicated
 to the Public Domain using the Creative Commons CC0 1.0 Universal (CC0 1.0)
-Public Domain Dedication: http://creativecommons.org/publicdomain/zero/1.0/</text>
\ No newline at end of file
+Public Domain Dedication: http://creativecommons.org/publicdomain/zero/1.0/</text>
+LicenseName: Other Copyleft Licenses
+LicenseComment: <text>See details at https://github.com/nexB/scancode-toolkit/blob/develop/src/licensedcode/data/licenses/other-copyleft.yml
```
Comment (pombredanne) [inline tests/formattedcode/data/spdx/license_ref/expected_with_text.tv]: Same as above. This is incorrect

File: tests/formattedcode/data/spdx/unicode/expected.tv [hunk 17]
```diff
@@ -75,4 +73,7 @@ ExtractedText: <text> * SOFTWARE LICENSE
 * ON ANY THEORY OF LIABILITY, INCLUDING, BUT NOT LIMITED TO, CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
-* DAMAGE.</text>
\ No newline at end of file
+* DAMAGE.</text>
+LicenseName: Agere BSD
```
Comment (pombredanne) [inline tests/formattedcode/data/spdx/unicode/expected.tv:77]: Let's not reorder these

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks!
See some comments for your review.

File: setup-mini.cfg [hunk 18]
```diff
@@ -105,7 +105,7 @@ install_requires =
     pymaven_patch >= 0.2.8
     requests >= 2.7.0
     saneyaml >= 0.6.0
-    spdx_tools == 0.7.0rc0
+    spdx_tools == 0.8.0rc2
```
Comment (AyanSinhaMahapatra) [inline setup-mini.cfg]: Thanks for the PR.
I had a question on this being a release-candidate.

The readme says the following:
```
The main features of v0.8 are:

    experimental support for the upcoming SPDX v3 specification (note, however, that support is neither complete nor stable at this point, as the spec is still evolving)
```

SPDX v3 is not out yet and the spec could change/evolve more from this point, and
it's admittedly `neither complete nor stable`, shouldn't we wait for a stable release of 
`spdx-tools` then?
(which I'm assuming would be after SPDX v3 would be out officially) 
Does this PR change the spec version of the SPDX output produced?

Comment (armintaenzertng) [reply setup-mini.cfg]: Thanks for voicing your concerns, @AyanSinhaMahapatra, but I can alleviate them! :)
This PR does not introduce any SPDX v3 features into scancode and makes no use of the experimental/unstable SPDX-3 code.
We will release 0.8.0 soon and I'll update the dependency then. The 0.8.0 version is stable for SPDX v2, which is the only part of the library used in scancode.

So, to answer your last question:
No, this PR does not change the spec version of the SPDX output produced.

File: tests/formattedcode/data/spdx/license_known/expected.tv (hunk 7 above)
Comment (armintaenzertng) [reply tests/formattedcode/data/spdx/license_known/expected.tv:1]: Regarding this and similar questions about the change of output:
During the refactoring of the python-tools we had to rewrite most of the serializers. For this, we followed the practices employed by the [official SPDX spec examples](https://github.com/spdx/spdx-spec/blob/development/v2.3.1/examples/SPDXTagExample-v2.3.spdx), especially the ordering of information and formatting of comments (which use double ## and "Information" in most places).

File: src/formattedcode/output_spdx.py (hunk 4 above)
Comment (armintaenzertng) [reply src/formattedcode/output_spdx.py]: This is the license information found in a single file. Information has no plural. It is also the official name in serializations, see [here](https://spdx.github.io/spdx-spec/v2.3/file-information/#86-license-information-in-file-field).

File: tests/formattedcode/data/spdx/license_known/expected.tv (hunk 9 above)
Comment (armintaenzertng) [reply tests/formattedcode/data/spdx/license_known/expected.tv:6]: The LicenseListVersion is getting "cleaned" in the process of regenerating test files, see [here](https://github.com/nexB/scancode-toolkit/blob/200fb10eb22218c36f68e37f2f9dc86342adaa35/tests/formattedcode/test_output_spdx.py#L151).
 I suppose the previous test file has not been regenerated since the cleaning code was introduced.

File: tests/formattedcode/data/spdx/license_known/expected.rdf (hunk 12 above)
Comment (armintaenzertng) [reply tests/formattedcode/data/spdx/license_known/expected.rdf:97]: There was a bug in the previous spdx-tools version that did not include the correct namespace in the SpdxIds. Now, as the document namespace includes a generated UUID, it can't be used for direct comparison.  Therefore I had to introduce the code [here](https://github.com/nexB/scancode-toolkit/pull/3456/files/26f26c5f036f0e3468d5552788216a55f042907b#diff-5d7d79946ca5faabcbfa18aead7e79acc42cb26fd70a254dd0142231d1d211e6R34-R36) to strip the namespace from the data.
Still, this will at least check that all SpdxIds have the same document namespace as they wouldn't be stripped otherwise.

Comment (armintaenzertng) [reply tests/formattedcode/data/spdx/license_known/expected.rdf:97]: Sorry, I misunderstood your comment.
The "SPDXRef-001" is not missing, it [can be found a good deal further up](https://github.com/nexB/scancode-toolkit/pull/3456/files/26f26c5f036f0e3468d5552788216a55f042907b?diff=split&w=0#diff-bcb5f143ba2af18266072e7c8004418d08e8b57400f643fa8c623cafde1e4430R17) with the package that the relationship belongs to. This is similar to the [RDF relationship example in the spec](https://spdx.github.io/spdx-spec/v2.3/relationships-between-SPDX-elements/#1113-examples).

File: tests/formattedcode/data/spdx/license_ref/expected_with_text.tv (hunk 13 above)
Comment (armintaenzertng) [reply tests/formattedcode/data/spdx/license_ref/expected_with_text.tv:88]: The order in the [official spec example](https://github.com/spdx/spdx-spec/blob/development/v2.3.1/examples/SPDXTagExample-v2.3.spdx#L292) and implemented in the serializer is: Id, extracted text, name, cross reference, comment.
So, not different from your proposal?

File: tests/formattedcode/data/spdx/license_ref/expected_with_text.tv (hunk 14 above)
Comment (armintaenzertng) [reply tests/formattedcode/data/spdx/license_ref/expected_with_text.tv:89]: The order is: Id first, then name after that. So the name that belongs to this Id is "Other Permissive License".
We should probably include an empty line before the Id to better separate the different licenses.

File: tests/formattedcode/data/spdx/license_ref/expected_with_text.tv (hunk 16 above)
Comment (armintaenzertng) [reply tests/formattedcode/data/spdx/license_ref/expected_with_text.tv]: As you reference the LicenseComment here, I'm not sure I know what you mean.
In case you meant the LicenseName, that corresponds to the Id "LicenseRef-scancode-other-copyleft", which seems fine to me.

File: tests/formattedcode/data/spdx/license_ref/expected_with_text.tv (hunk 14 above)
Comment (armintaenzertng) [reply tests/formattedcode/data/spdx/license_ref/expected_with_text.tv:89]: Ah, actually, the empty lines get stripped during the cleaning step.

File: src/formattedcode/output_spdx.py (hunk 1 above)
Comment (armintaenzertng) [reply src/formattedcode/output_spdx.py]: Sure, I fixed that.

File: src/formattedcode/output_spdx.py (hunk 2 above)
Comment (armintaenzertng) [reply src/formattedcode/output_spdx.py]: done

File: src/formattedcode/output_spdx.py (hunk 5 above)
Comment (armintaenzertng) [reply src/formattedcode/output_spdx.py]: The append function circumvented our previous type checking, so I was in the habit of avoiding that. Shouldn't be an issue anymore, I changed all occurrences to `.append()`.

File: src/formattedcode/output_spdx.py (hunk 6 above)
Comment (armintaenzertng) [reply src/formattedcode/output_spdx.py:369]: As this issue was not introduced by this PR, I opened #3457 for it.

File: src/formattedcode/output_spdx.py (hunk 3 above)
Comment (armintaenzertng) [reply src/formattedcode/output_spdx.py:289]: Correct me if I'm wrong, but isn't this what happens here? If we have a LicenseRef-, i.e. not an SPDX license, we enter the if clause and get the matched text.
This behavior is the same as in the previous implementation, so if you think this is a problem, I'd like to open a separate issue for this as this is not really part of this PR.

File: setup-mini.cfg (hunk 18 above)
Comment (AyanSinhaMahapatra) [reply setup-mini.cfg]: Thanks for the clarifications @armintaenzertng :smile:

Comment (mjherzog) [reply setup-mini.cfg]: @armintaenzertng Would it be possible to correct the README for v0.08 to reflect what you explained?

Comment (armintaenzertng) [reply setup-mini.cfg]: The README was updated to hopefully make clear that the SPDX3 code is not used in production yet.

**Comment by armintaenzertng** [comment]
@pombredanne, I believe I addressed all your comments above (apart from the reordering, for which we will continue to adhere to the specification examples) and implemented them or opened an issue if that problem already existed prior to this PR.
Also, we have officially released spdx-tools v0.8.0, I updated the version in this PR.

The CI currently fails for `macos12` only. The failing test `test_scan_does_scan_php_composer` succeeds for all other platforms and also locally for me. This seems like an issue that's unrelated to this PR.

**Comment by AyanSinhaMahapatra** [comment]
> we have officially released spdx-tools v0.8.0, I updated the version in this PR.

Thanks @armintaenzertng !

> The CI currently fails for macos12 only. The failing test test_scan_does_scan_php_composer succeeds for all other platforms and also locally for me. This seems like an issue that's unrelated to this PR.

These are flukes we see sometimes, I've reran the tests and all green now.

**Comment by maxhbr** [comment]
Hey @pombredanne and @AyanSinhaMahapatra , just was checking in, if I can help to get it merged. Are there still open concerns?

File: src/formattedcode/output_spdx.py [hunk 19]
```diff
@@ -63,50 +67,6 @@ def logger_debug(*args):
 Output plugins to write scan results in SPDX format.
 """
 
-_spdx_list_is_patched = False
```
Comment (pombredanne) [inline src/formattedcode/output_spdx.py:66]: Why did you remove this?

**Review by pombredanne — CHANGES_REQUESTED** [review]

File: src/formattedcode/output_spdx.py (hunk 19 above)
Comment (maxhbr) [reply src/formattedcode/output_spdx.py:66]: The `LICENSE_MAP` was dropped in https://github.com/spdx/tools-python/commit/cb6be81c9f0705120f3565989a7e7d5fc585e129, during the refactoring. The interface used for this patching is no longer available. As the spdx-tools are now dependent on https://github.com/nexB/license-expression, maybe there is another way of patching?

File: src/formattedcode/output_spdx.py (hunk 3 above)
Comment (pombredanne) [reply src/formattedcode/output_spdx.py:289]: Ack. Thanks for the clarification

File: src/formattedcode/output_spdx.py (hunk 4 above)
Comment (pombredanne) [reply src/formattedcode/output_spdx.py]: if this is a collection, I would rather have a plural as a convention irrespective of how things may be named in SPDX.

File: src/formattedcode/output_spdx.py (hunk 19 above)
Comment (pombredanne) [reply src/formattedcode/output_spdx.py:66]: @maxhbr Thanks for the clarification.... we need to find a way as ScanCode cannot rely on anything else but the licenses it knows about in a given version, i.e. the licenses must be the ones available in ScanCode toolkit not these provided by any other library (not even license-expression)

**Review by pombredanne — CHANGES_REQUESTED** [review]

File: src/formattedcode/output_spdx.py (hunk 19 above)
Comment (armintaenzertng) [reply src/formattedcode/output_spdx.py:66]: @pombredanne I believe this is not an issue. The only place in the old/current code that uses the `LICENSE_MAP` is [this](https://github.com/nexB/scancode-toolkit/blob/5dfcfcbbcda9a5bcd70a1fcf887babf661a87a03/src/formattedcode/output_spdx.py#L316), i.e. the construction of a `License` object. In the new 0.8 of the python-tools we don't have `License` objects anymore. Instead, every license is just a `LicenseExpression` object which carries no additional information about SPDX or otherwise for the purpose of serialization that we discuss here.
The logic that determines the correct SPDX license identifier happens [in these lines](https://github.com/armintaenzertng/scancode-toolkit/blob/update-spdx-tools-version/src/formattedcode/output_spdx.py#L278-L284) which are entirely ScanCode-internal. The only external non-trivial thing happening after that is [this line](https://github.com/armintaenzertng/scancode-toolkit/blob/4918b04efaf4ffa0b9c446787baa0029ce00093a/src/formattedcode/output_spdx.py#L286), which just converts the `spdx_id` string to a `LicenseExpression` object, but without any "license list logic" behind it, especially as the `licensing` is defined via `Licensing()` and not via `get_spdx_licensing()`. That `LicenseExpression` is in this case no more than a glorified string that gets added to the lists of its file/package and subsequently serialized to a simple string in the final output.

Please correct me if I'm wrong, but I see no need to keep the code fragment in question, as we make no use of ScanCode-external license lists.

**Comment by maxhbr** [comment]
@pombredanne , does https://github.com/nexB/scancode-toolkit/pull/3456#discussion_r1300071128 resolve the issue?

**Comment by armintaenzertng** [comment]
Just a small update: I rebased the code on the current state of the repo and updated the spdx-tools dependency to 0.8.1 (the update has no impact on the code in this PR).

**Comment by armintaenzertng** [comment]
Hi @AyanSinhaMahapatra,
could you please have a look at https://github.com/nexB/scancode-toolkit/pull/3456#discussion_r1300071128 and see if that solves the open issue with this PR?

**Comment by maxhbr** [comment]
hey, any updates?

File: tests/formattedcode/data/spdx/license_known/expected.tv (hunk 7 above)
Comment (pombredanne) [reply tests/formattedcode/data/spdx/license_known/expected.tv:1]: It does not make any sense: ## makes the file bigger for no good reason. The fact that spec examples use this weird way does not make it part of the spec FWIW. The spec states that a # is a comment not that you should use a double ##. Now this is not critical so I can live with this change but this feels like cargo culting.

Also the logical order from most significant to least significant is namespace then name. This is BTW also the order of the spec examples you reference https://github.com/spdx/spdx-spec/blob/development/v2.3.1/examples/SPDXTagExample-v2.3.spdx#L3 

I do not care too much for the change so I can live with it  but there is no rationale for this change that I can understand.

**Review by pombredanne — APPROVED** [review]
Thanks... I have added a new nit for your review wrt. the field ordering problem but I am goimng to merge this as-is now.
